controllers/elasticdl/elasticdljob_controller.go

```diff
@@ -0,0 +1,206 @@
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch owner resource with create event filter.
+	if err = c.Watch(&source.Kind{Type: &elasticdlv1alpha1.ElasticDLJob{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc: onOwnerCreateFunc(r),
+	}); err != nil {
+		return err
+	}
+
+	// Watch managed resource with owner and create/delete event filter.
+	if err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		OwnerType:    &elasticdlv1alpha1.ElasticDLJob{},
+		IsController: true,
+	}, predicate.Funcs{
+		CreateFunc: r.ctrl.OnPodCreateFunc,
+		UpdateFunc: r.ctrl.OnPodUpdateFunc,
+		DeleteFunc: r.ctrl.OnPodDeleteFunc,
+	}); err != nil {
+		return err
+	}
+
+	return c.Watch(&source.Kind{Type: &corev1.Service{}}, &handler.EnqueueRequestForOwner{
+		OwnerType:    &elasticdlv1alpha1.ElasticDLJob{},
+		IsController: true,
+	}, predicate.Funcs{
+		CreateFunc: r.ctrl.OnServiceCreateFunc,
+		UpdateFunc: r.ctrl.OnServiceUpdateFunc,
+		DeleteFunc: r.ctrl.OnServiceDeleteFunc,
+	})
+}
+
+func (r *ElasticDLJobReconciler) ControllerName() string {
+	return controllerName
+}
+
+// GetAPIGroupVersionKind returns the GroupVersionKind of the API
+func (r *ElasticDLJobReconciler) GetAPIGroupVersionKind() schema.GroupVersionKind {
+	return elasticdlv1alpha1.GroupVersionKind
+}
+
+// GetAPIGroupVersion returns the GroupVersion of the API
+func (r *ElasticDLJobReconciler) GetAPIGroupVersion() schema.GroupVersion {
+	return elasticdlv1alpha1.SchemeGroupVersion
+}
+
+// GetGroupNameLabelValue returns the Group Name(value) in the labels of the job
+func (r *ElasticDLJobReconciler) GetGroupNameLabelValue() string {
+	return elasticdlv1alpha1.GroupName
+}
+
+// GetDefaultContainerName returns the default container name in pod
+func (r *ElasticDLJobReconciler) GetDefaultContainerName() string {
+	return elasticdlv1alpha1.DefaultContainerName
+}
+
+// GetDefaultContainerPortName Get the default container port name
+func (r *ElasticDLJobReconciler) GetDefaultContainerPortName() string {
+	return elasticdlv1alpha1.DefaultPortName
+}
+
+// GetDefaultContainerPortNumber get the default container port number
+func (r *ElasticDLJobReconciler) GetDefaultContainerPortNumber() int32 {
+	return elasticdlv1alpha1.DefaultPort
+}
+
+func (r *ElasticDLJobReconciler) GetReconcileOrders() []v1.ReplicaType {
+	return []v1.ReplicaType{
+		elasticdlv1alpha1.ElasticDLReplicaTypeMaster,
+	}
+}
+
+func (r *ElasticDLJobReconciler) IsMasterRole(replicas map[v1.ReplicaType]*v1.ReplicaSpec, rtype v1.ReplicaType, index int) bool {
+	_, ok := replicas[elasticdlv1alpha1.ElasticDLReplicaTypeMaster]
+	return ok && rtype == elasticdlv1alpha1.ElasticDLReplicaTypeMaster
+}
+
+// SetClusterSpec sets the cluster spec for the pod
+func (r *ElasticDLJobReconciler) SetClusterSpec(job interface{}, podTemplate *corev1.PodTemplateSpec, rtype, index string) error {
+	_, ok := job.(*elasticdlv1alpha1.ElasticDLJob)
```

Comment:
just return nil. No need to verify job and index

Reply:
Done